### PR TITLE
fix file_sd never stop update 'custom_sd.json' file in adapter.go

### PR DESCRIPTION
**Pull request by earthdiaosi**
fixed: #4541

### CODE REVIEW

File: documentation/examples/custom-sd/adapter/adapter.go [hunk 1]
```diff
@@ -34,6 +36,31 @@ type customSD struct {
 	Labels  map[string]string `json:"labels"`
 }
 
+//return targetsHash.Sum64() ^ labelsHash.Sum64(), nil
+func (sd *customSD) hash() (uint64, error) {
```
Comment (simonpasquier) [inline documentation/examples/custom-sd/adapter/adapter.go]: I think you could leverage https://godoc.org/github.com/prometheus/common/model#LabelSet.Fingerprint which does mostly the same thing.

File: documentation/examples/custom-sd/adapter/adapter.go [hunk 2]
```diff
@@ -72,12 +99,22 @@ func (a *Adapter) generateTargetGroups(allTargetGroups map[string][]*targetgroup
...
 			}
-			// Make a unique key, including the current index, in case the sd_type (map key) and group.Source is not unique.
-			key := fmt.Sprintf("%s:%s:%d", k, group.Source, i)
-			tempGroups[key] = &customSD{
+
+			sdGroup := customSD{
 				Targets: newTargets,
 				Labels:  newLabels,
 			}
+			// Make a unique key, including the current index, in case the sd_type (map key) and group.Source is not unique.
```
Comment (cstyan) [inline documentation/examples/custom-sd/adapter/adapter.go]: this comment should be updated based on your changes

File: documentation/examples/custom-sd/adapter/adapter.go (hunk 1 above)
Comment (cstyan) [reply documentation/examples/custom-sd/adapter/adapter.go]: Yep, for a `targetgroup` targets is `[]model.LabelSet` and labels are `model.LabelSet`

https://github.com/prometheus/prometheus/blob/master/discovery/targetgroup/targetgroup.go#L23-L33

File: documentation/examples/custom-sd/adapter/adapter.go [hunk 3]
```diff
@@ -34,6 +36,31 @@ type customSD struct {
 	Labels  map[string]string `json:"labels"`
 }
 
+//return targetsHash.Sum64() ^ labelsHash.Sum64(), nil
```
Comment (cstyan) [inline documentation/examples/custom-sd/adapter/adapter.go]: If you want to keep these commented lines, comments should be in the form of a full sentence. `hash turns both the targets and labels for sd into hashes, and then returns the targetHash ^ labelsHash.` as an example.

File: documentation/examples/custom-sd/adapter/adapter.go (hunk 1 above)
Comment (earthdiaosi) [reply documentation/examples/custom-sd/adapter/adapter.go]: done

File: documentation/examples/custom-sd/adapter/adapter.go (hunk 2 above)
Comment (earthdiaosi) [reply documentation/examples/custom-sd/adapter/adapter.go]: done

**Comment by earthdiaosi** [comment]
see commit: https://github.com/prometheus/prometheus/pull/4567/commits/1b5910778a317cee86d59e38ff706b549f268bd5

File: documentation/examples/custom-sd/adapter/adapter.go [hunk 4]
```diff
@@ -72,12 +82,17 @@ func (a *Adapter) generateTargetGroups(allTargetGroups map[string][]*targetgroup
...
+
+			sdGroup := customSD{
+
 				Targets: newTargets,
 				Labels:  newLabels,
 			}
+			// Make a unique key, including group's fingerprint, in case the sd_type (map key) and group.Source is not unique.
+			groupFingerprint := fingerprint(group)
+			key := fmt.Sprintf("%s:%s:%s", k, group.Source, groupFingerprint.String())
+			fmt.Println(key)
```
Comment (simonpasquier) [inline documentation/examples/custom-sd/adapter/adapter.go]: to be removed

Comment (earthdiaosi) [reply documentation/examples/custom-sd/adapter/adapter.go]: done

**Comment by earthdiaosi** [comment]
see commit: https://github.com/prometheus/prometheus/pull/4567/commits/c19005041208cef68b5fbec6b21d37f4adf5bda9

**Comment by cstyan** [comment]
@simonpasquier

**Comment by simonpasquier** [comment]
Can you rebase?

**Comment by cstyan** [comment]
I think you removed some stuff that was in master when rebasing and fixing conflicts. See the failing tests.

**Comment by simonpasquier** [comment]
@earthdiaosi would you have time to fix the tests?

**Comment by earthdiaosi** [comment]
@simonpasquier @cstyan fixed.

**Comment by cstyan** [comment]
:+1: lgtm, thanks @earthdiaosi

**Comment by earthdiaosi** [comment]
when shall merge this pull request @simonpasquier @cstyan

**Comment by simonpasquier** [comment]
Sorry I thought I already hit the merge button. Thanks for the work!

**Comment by earthdiaosi** [comment]
Thank you very much and I'm glad to be a Contributor.  thanks @simonpasquier @cstyan

**Pull request merged**